### ISSUE TITLE
How to create a single page site? (Solved)

### ISSUE THREAD

**Issue by voorjaar** [TypeStrong/typedoc#2096]
## Search terms

single page, spa, mpa, custom plugin, docs,

## Question

How to generate an SPA using typedoc, or making an MPA works like an SPA?

### Why

SPA has a better user experience, such as no need to refresh the page, no loss of scroll position. And you can add page switching animation and progress bar, etc.

## Answer

I have created a plugin called [typedoc-plugin-mtos](https://github.com/voorjaar/typedoc-plugin-mtos) for typedoc to implement this functionality.  It's based on [mtos](https://github.com/voorjaar/mtos).

This issue was created to help people who may have similar questions and provide a solution.

> Checkout [this page](https://voorjaar.github.io/typedoc-plugin-mtos/) for a preview demo.

### How it works?

The [mtos](https://github.com/voorjaar/mtos) works similar to SPA, but is based on native DOM. The workflow like this:

1. Query all `<a>` elements which property `href` includes current host.
2. Add a `onclick` function, when clicked, fetch the HTML content from target link.
3. Push link to [history](https://developer.mozilla.org/en-US/docs/Web/API/History_API) state.
4. Diff current document with fetched content, update changed elements.
5. Goto 1.

So basically you still serve the static HTML files, but the user experience is the same as SPA with incremental requests via fetch API. It's still an MPA, but it works like an SPA in client side.

**Comment by Gerrit0** [comment]
Neat! It's showing up on the plugins page now too: https://typedoc.org/guides/plugins/
